### PR TITLE
planner,util: optimize encode plan performance by avoid fmt.Sprintf/Fprintf

### PR DESCRIPTION
**Pull request by ClSlaid**
### What problem does this PR solve?

Problem Summary:
- encodePlan() consumes too much time.

### What is changed and how it works?

What's Changed: 
replaced fmt.Fprintf into strings.Builder

How it Works:
Use strings.Builder to avoid string truncating and formating.

### Improvement

Go Bench

```go
// adding following benchmark code into tidb/planner/core/plan-test.go
func (s *testPlanNormalize) BenchmarkEncodePointGetPlan(c *C) {
	tk := testkit.NewTestKit(c, s.store)
	tk.MustExec("use test")
	tk.MustExec("drop table if exists th")
	tk.MustExec("create table th (a bigint key, b int);")

	query := "select * from th where a = 1"
	tk.Se.GetSessionVars().PlanID = 0
	tk.MustExec(query)
	info := tk.Se.ShowProcess()
	c.Assert(info, NotNil)
	p, ok := info.Plan.(core.PhysicalPlan)
	c.Assert(ok, IsTrue)
	c.ResetTimer()
	for i := 0; i < c.N; i++ {
		core.EncodePlan(p)
	}
}
```

Go Bench outcome in tidb/planner/core:
```zsh
go test -check.b=true -check.btime=5s | ripgrep PASS # this PR
# PASS: plan_test.go:539: testPlanNormalize.BenchmarkDecodePlan        100          60852026 ns/op
# PASS: plan_test.go:571: testPlanNormalize.BenchmarkEncodePlan         20         416020938 ns/op
# PASS: plan_test.go:594: testPlanNormalize.BenchmarkEncodePointGetPlan   10000000              1104 ns/op
go test -check.b=true -check.btime=5s | ripgrep PASS # master
# PASS: plan_test.go:538: testPlanNormalize.BenchmarkDecodePlan        100          60730623 ns/op
# PASS: plan_test.go:570: testPlanNormalize.BenchmarkEncodePlan         20         435428859 ns/op
# PASS: plan_test.go:592: testPlanNormalize.BenchmarkEncodePointGetPlan    5000000              1816 ns/op
```
We can conclude that This PR significantly improved the performance of function getEncodePlan().

pprof diagrams:
- master:
图片

- this PR:
图片

This PR should reduces time comsumed by `func (*planEncoder) encodePlan()` by 6.5%.

#### Point Get Improvement

| TiDB Version | This PR | This PR | *master | master |
| :------: | :----: | :----: | :----: | :----: |
| @@global.tidb_stmt_summary_max_stmt_count | 3000 | 1 | 3000 | 1 |
| sysbench average QPS | 63052 | 28869 | 62658 | 28453 |

slightly improved sysbench benchmark.

### Check List 

Tests 

- [x] Unit test
- [x] Integration test

Side effects

  Used strings.Builder, which is Go 1.10+ only

### Release note 

- no release note.

### CODE REVIEW

**Comment by ClSlaid** [comment]
/cc @crazycs520 @breeswish 
PTAL.

File: planner/core/point_get_plan.go [hunk 1]
```diff
@@ -166,23 +168,30 @@ func (p *PointGetPlan) AccessObject(normalized bool) string {
 
 // OperatorInfo implements dataAccesser interface.
 func (p *PointGetPlan) OperatorInfo(normalized bool) string {
-	buffer := bytes.NewBufferString("")
+	var buffer strings.Builder
```
Comment (breeswish) [inline planner/core/point_get_plan.go:171]: Can we do the same thing to other `OperatorInfo`?

File: planner/core/point_get_plan.go [hunk 2]
```diff
@@ -166,23 +168,30 @@ func (p *PointGetPlan) AccessObject(normalized bool) string {
...
+			buffer.WriteString("handle:?")
 		} else {
 			if p.UnsignedHandle {
-				fmt.Fprintf(buffer, "handle:%d, ", uint64(p.Handle.IntValue()))
+				// fmt.Fprintf(buffer, "handle:%d, ", uint64(p.Handle.IntValue()))
+				buffer.WriteString("handle:" + strconv.FormatUint(uint64(p.Handle.IntValue()), 10))
 			} else {
-				fmt.Fprintf(buffer, "handle:%s, ", p.Handle)
+				// fmt.Fprintf(buffer, "handle:%s, ", p.Handle)
+				buffer.WriteString("handle:" + p.Handle.String())
```
Comment (breeswish) [inline planner/core/point_get_plan.go:185]: There is string concat here. I guess simply using `buffer.WriteString` × 2 will be more efficient.

Comment (ClSlaid) [reply planner/core/point_get_plan.go:185]: ok, sorry I forgot deleting old code...

File: planner/core/point_get_plan.go (hunk 1 above)
Comment (ClSlaid) [reply planner/core/point_get_plan.go:171]: If our buffer is or will be used as string, it's better to switch to `strings.Builder`.

File: planner/core/plan_test.go [hunk 3]
```diff
@@ -571,9 +572,7 @@ func (s *testPlanNormalize) BenchmarkEncodePlan(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
 	tk.MustExec("use test")
 	tk.MustExec("drop table if exists th")
-	tk.MustExec("set @@session.tidb_enable_table_partition = 1")
-	tk.MustExec(`set @@tidb_partition_prune_mode='` + string(variable.Static) + `'`)
-	tk.MustExec("create table th (i int, a int,b int, c int, index (a)) partition by hash (a) partitions 8192;")
+	tk.MustExec("create table th (i int primary key);")
```
Comment (crazycs520) [inline planner/core/plan_test.go:575]: No need to change this function?

File: planner/core/plan_test.go [hunk 4]
```diff
@@ -589,3 +588,27 @@ func (s *testPlanNormalize) BenchmarkEncodePlan(c *C) {
 		core.EncodePlan(p)
 	}
 }
+
+func (s *testPlanNormalize) BenchmarkEncodePlanPProf(c *C) {
+	go func() {
+		fmt.Println(http.ListenAndServe("localhost:8086", nil))
+	}()
```
Comment (crazycs520) [inline planner/core/plan_test.go:595]: This is only use for debug, please remove this.

File: planner/core/plan_test.go [hunk 5]
```diff
@@ -589,3 +588,27 @@ func (s *testPlanNormalize) BenchmarkEncodePlan(c *C) {
...
+	query := "select * from th where i = 2333;"
+	tk.Se.GetSessionVars().PlanID = 0
+	tk.MustExec(query)
+	info := tk.Se.ShowProcess()
+	c.Assert(info, NotNil)
+	p, ok := info.Plan.(core.PhysicalPlan)
+	c.Assert(ok, IsTrue)
+	tk.Se.GetSessionVars().StmtCtx.RuntimeStatsColl = nil
+	c.ResetTimer()
+	for {
```
Comment (crazycs520) [inline planner/core/plan_test.go:611]: Better change to `for i := 0; i < c.N; i++ {`

Dead loop only use for debug.

File: planner/core/plan_test.go [hunk 6]
```diff
@@ -589,3 +588,27 @@ func (s *testPlanNormalize) BenchmarkEncodePlan(c *C) {
...
+	tk.MustExec("set @@tidb_slow_log_threshold=200000")
+
+	query := "select * from th where i = 2333;"
+	tk.Se.GetSessionVars().PlanID = 0
+	tk.MustExec(query)
+	info := tk.Se.ShowProcess()
+	c.Assert(info, NotNil)
+	p, ok := info.Plan.(core.PhysicalPlan)
+	c.Assert(ok, IsTrue)
+	tk.Se.GetSessionVars().StmtCtx.RuntimeStatsColl = nil
```
Comment (crazycs520) [inline planner/core/plan_test.go:609]: You can remove this line to found more place need optimize.

File: planner/core/point_get_plan.go [hunk 7]
```diff
@@ -356,21 +368,24 @@ func (p *BatchPointGetPlan) AccessObject(_ bool) string {
...
 func (p *BatchPointGetPlan) OperatorInfo(normalized bool) string {
-	buffer := bytes.NewBufferString("")
+	var buffer strings.Builder
 	if p.IndexInfo == nil {
 		if normalized {
-			fmt.Fprintf(buffer, "handle:?, ")
+			buffer.WriteString("handle:?, ")
 		} else {
-			fmt.Fprintf(buffer, "handle:%v, ", p.Handles)
+			//fmt.Fprintf(buffer.WriteString("handle:%v, ", p.Handles)
```
Comment (crazycs520) [inline planner/core/point_get_plan.go:376]: remove this line?

File: planner/core/point_get_plan.go [hunk 8]
```diff
@@ -356,21 +368,24 @@ func (p *BatchPointGetPlan) AccessObject(_ bool) string {
...
+	buffer.WriteString("keep order:")
+	buffer.WriteString(strconv.FormatBool(p.KeepOrder))
+	buffer.WriteString(", desc:")
+	buffer.WriteString(strconv.FormatBool(p.Desc))
 	if p.Lock {
-		fmt.Fprintf(buffer, "lock, ")
-	}
-	if buffer.Len() >= 2 {
-		buffer.Truncate(buffer.Len() - 2)
+		// fmt.Fprintf(buffer, "lock, ")
```
Comment (crazycs520) [inline planner/core/point_get_plan.go:387]: ditto.

**Comment by breeswish** [comment]
Great work! Could you provide benchmark results about how it performs now at eviction?

**Comment by ClSlaid** [comment]
/run-all-tests

> Great work! Could you provide benchmark results about how it performs now at eviction?

added

File: planner/core/explain.go [hunk 9]
```diff
@@ -225,46 +245,53 @@ func (p *PhysicalTableScan) AccessObject(normalized bool) string {
...
+			buffer.WriteString(")")
 		default:
-			fmtStr = "%s, "
+			buffer.WriteString(pkCol.ExplainInfo())
+			buffer.WriteString(", ")
 		}
-		fmt.Fprintf(buffer, fmtStr, pkCol.ExplainInfo())
 	}
 	if len(p.rangeDecidedBy) > 0 {
+		// TODO: optimize Fprintf
```
Comment (crazycs520) [inline planner/core/explain.go:262]: use `strings.Builder` and then remove this todo?

File: planner/core/explain.go [hunk 10]
```diff
@@ -393,11 +427,12 @@ func (p *PhysicalIndexLookUpReader) accessObject(sctx sessionctx.Context) string
 		return ""
 	}
 
-	var buffer bytes.Buffer
+	var buffer strings.Builder
```
Comment (crazycs520) [inline planner/core/explain.go:430]: move after line433.

File: planner/core/plan_test.go [hunk 11]
```diff
@@ -16,6 +16,7 @@ package core_test
 import (
 	"bytes"
 	"fmt"
+	_ "net/http/pprof"
```
Comment (crazycs520) [inline planner/core/plan_test.go:19]:
```suggestion
```

File: planner/core/point_get_plan.go [hunk 12]
```diff
@@ -133,21 +133,26 @@ func (p *PointGetPlan) ExplainNormalizedInfo() string {
 
 // AccessObject implements dataAccesser interface.
 func (p *PointGetPlan) AccessObject(normalized bool) string {
-	buffer := bytes.NewBufferString("")
+	var buffer strings.Builder
 	tblName := p.TblInfo.Name.O
-	fmt.Fprintf(buffer, "table:%s", tblName)
+	buffer.WriteString("table:" + tblName)
```
Comment (crazycs520) [inline planner/core/point_get_plan.go:138]:
```suggestion
	buffer.WriteString("table:")
	buffer.WriteString(tblName)
```

File: planner/core/point_get_plan.go (hunk 1 above)
Comment (crazycs520) [reply planner/core/point_get_plan.go:174]: move to line 177.

File: planner/core/point_get_plan.go [hunk 13]
```diff
@@ -166,23 +172,28 @@ func (p *PointGetPlan) AccessObject(normalized bool) string {
 
 // OperatorInfo implements dataAccesser interface.
 func (p *PointGetPlan) OperatorInfo(normalized bool) string {
-	buffer := bytes.NewBufferString("")
+	var buffer strings.Builder
+	if p.Handle == nil && !p.Lock {
+		return ""
+	}
```
Comment (crazycs520) [inline planner/core/point_get_plan.go:178]:
```suggestion
	if p.Handle == nil && !p.Lock {
		return ""
	}
	var buffer strings.Builder
```

**Review by crazycs520 — COMMENTED** [review]
REST LGTM

**Review by crazycs520 — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**